docs/oprf.typst

```diff
@@ -172,14 +139,329 @@ One of the main use-cases of BabyJubJub is to build a digital signature scheme f
+
+== Full Distributed OPRF-Based Nullifier Protocol
+
+We give the full verifiable OPRF based distributed nullifier service construction in @fig:full_protocol. For the description of the zero knowledge proofs $pi_1$ and $pi_2$ we refer to @sec:ZK.
+
+#figure(
+  rect(table(
+    columns: (2fr, 1fr, 2fr),
+    stroke: none,
+    table.header([Client($sk, pk, uid, rpid, action, K$)], [], [$n$ Server($k_i, K = k dot B$)]),
+    table.hline(stroke: 0.05em),
+    [$beta arrowrl bb(Z)_q$], [], [],
+    [$q <- H'(uid, rpid, action)$], [], [],
+    [$sigma <-$`Sign`$(sk, q)$], [], [],
+    [$A <- beta dot H(q)$], [], [],
+    [$pi_1 <-$`prove`$(sigma, A,$ `valid`$(pk))$],
+    math.attach($arrow.r$, t: $A, pi_1, action, rpid$),
+    [if `verify`$(pi_1, A, action, rpid) = bot$ then abort],
+    [], [], [$(r_i, C_i, R_(i,1), R_(i,2)) <-$ `dlog.partial_commitments`$(k_i, A, B)$],
+    [$C <-$`Reconstruct`$([C])$], math.attach($arrow.l$, t: $[C], [R_1], [R_2]$), [],
+    [$R_1 <-$`Reconstruct`$([R_1])$], [], [],
+    [$R_2 <-$`Reconstruct`$([R_2])$], [], [],
+    [$e <- H(K, A, C, B,R_1, R_2) in bb(F)_p$], math.attach($arrow.r$, t: $e$), [$s_i <- r_i + e dot k_i$],
+    [], math.attach($arrow.l$, t: $[s]$), [],
+    [$s <-$`Reconstruct`$([s])$], [], [],
+    [$n <- H'(q, (beta^(-1) dot C))$], [], [],
+    [$pi_2 <-$`prove`$(sigma, A,$`valid`$(pk),$ `dlog.verify`$(e,s), n)$], [], [],
+    [Output $(n, pi_2)$],
+  )),
+  caption: [The distributed verifiable TwoHashDH based nullifier construction derived from @fig:twohashdh.],
+  kind: "scheme",
+  supplement: [Scheme],
+)
+<fig:full_protocol>
+
+=== Clients Zero Knowledge Proofs <sec:ZK>
+
+We describe the ZK proofs $pi_1$ and $pi_2$ from @fig:full_protocol in this section.
 
+==== Query Proof $pi_1$
+
+The goal of the query proof $pi_1$ is to convince the server that a client is authorized to send a request. Therefore, `valid`$(pk)$ is a core part of this zero knowledge which shows that the used public key $pk$ is in some kind of allowlist. To prove knowledge of the corresponding private key $sk$ we opt to verify a signature $sigma <-$`Sign`$(sk, q)$ of the actual query $q$ inside the ZK proof. We use this signature to not have $sk$ as a private witness in the ZK proof. This has the advantage that a client can securely outsource proof generation to, e.g., an MPC-based prover network without having to secret share its key $sk$ with them, minimizing damage in case of a privacy breach.
+
+Finally, $pi_1$ binds everything together by proving the correct calculation of the query point $A$ from its part.
+
+In more details, proof $pi_1$ consists of the following statements:
+#figure(
+  box(stroke: black, inset: 1em)[
+    #align(left)[
```

Comment:
These two could use a Title/leading paragraph inside the box?
something formatted like `ProofName(inputs) -> outputs:` in bold or so, wdyt?